### ISSUE TITLE
Sample graceful shutdown code doesn't work with Cats Effect 3

### ISSUE THREAD

**Issue by fqshopify** [fd4s/fs2-kafka#994]
AFAICT the example shared on the [FS2 website](https://fd4s.github.io/fs2-kafka/docs/consumers#graceful-shutdown) only works with FS2 versions that are based on **Cats Effect 2** (and not with FS2 versions that are based on **Cats Effect 3**). 

# Example

Here is a [repo](https://github.com/fqshopify/fs2Graceful) with 2 commits on the `main` branch: 

1. [Cats Effect 2](https://github.com/fqshopify/fs2Graceful/commit/fcf3c33f6a02ae7842fdec11cc9705349dc34c17) commit
  When you run the unit test on this branch you will see the following expected behaviour: 
```
Starting processing of record at offset = 0
Finished processing of record at offset = 0
Starting processing of record at offset = 1
Finished processing of record at offset = 1
Starting processing of record at offset = 2
Finished processing of record at offset = 2
Starting processing of record at offset = 3
Cancelling task!
Starting graceful shutdown
Finished processing of record at offset = 3
Starting processing of record at offset = 4
Finished processing of record at offset = 4
Starting processing of record at offset = 5
Finished processing of record at offset = 5
Succeeded in graceful shutdown
```
As you can see, the app finishes processing the current record at offset = 3 and any records that were buffered internally (i.e. records at offset 4 & 5) before shutting down. Note there are more records past offset 5 that are never processed, as expected. 

2. [Cats Effect 3](https://github.com/fqshopify/fs2Graceful/commit/91f06264e99b394a2ea381e0aac105972f962c74) commit

When you run the same unit test with Cats Effect 3 versions of FS2, you see the following: 

```
Starting processing of record at offset = 0
Finished processing of record at offset = 0
Starting processing of record at offset = 1
Finished processing of record at offset = 1
Starting processing of record at offset = 2
Finished processing of record at offset = 2
Starting processing of record at offset = 3
Cancelling task!
Finished processing of record at offset = 3
Starting processing of record at offset = 4
Finished processing of record at offset = 4
Starting processing of record at offset = 5
Finished processing of record at offset = 5
Starting processing of record at offset = 6
Finished processing of record at offset = 6
Starting processing of record at offset = 7
Finished processing of record at offset = 7
Starting processing of record at offset = 8
Finished processing of record at offset = 8
Starting processing of record at offset = 9
Finished processing of record at offset = 9
Starting processing of record at offset = 10
Finished processing of record at offset = 10
Starting processing of record at offset = 11

java.util.concurrent.TimeoutException: Future timed out after [30 seconds]

	at scala.concurrent.impl.Promise$DefaultPromise.tryAwait0(Promise.scala:248)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:261)
	at scala.concurrent.Await$.$anonfun$result$1(package.scala:201)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:62)
	at scala.concurrent.Await$.result(package.scala:124)
	at munit.internal.PlatformCompat$.$anonfun$waitAtMost$1(PlatformCompat.scala:21)
	at scala.util.Try$.apply(Try.scala:210)
	at munit.internal.PlatformCompat$.waitAtMost(PlatformCompat.scala:21)
	at munit.FunSuite.waitForCompletion(FunSuite.scala:51)
	at munit.FunSuite.$anonfun$test$1(FunSuite.scala:37)
	at munit.MUnitRunner.$anonfun$runTestBody$1(MUnitRunner.scala:296)
```

The test eventually times out after 30 seconds as the app never gracefully shuts down (note the lack of a "Starting graceful shutdown" message in the output above). In addition, the console output shows the app continues consuming messages long past when the "Cancelling task" message appears. 

# Ask 

I would appreciate any pointers on how to make this work with Cats Effect version 3, and I would be happy to submit a PR to update the example on the website if successful.

**Comment by fqshopify** [comment]
cc: @bplommer

**Comment by fqshopify** [comment]
also tagging @LMnet who originally authored the feature in https://github.com/fd4s/fs2-kafka/pull/421. Would appreciate thoughts on how to "fix" this.

**Comment by LMnet** [comment]
Hi @fqshopify

Thank you for the issue reporting. I really appreciate that you created a repository with the reproduction steps. It really helps.

I checked your repo and yes, my example of graceful shutdown logic doesn't work with ce3. It's because of the chaged cancellation semantics on fibers. In ce2 finalization code runs concurrently wtth a fiber which was cancelled. On ce3 finalization code will run only after fiber full cancel. It means that a whole `release` block in `bracketCase` will not run until `use` block is not cancelled. But the code inside `use` block marked as `uncancellable`. And it leads to a lock.

Fortunately, it doens't mean that there is no way to use `stopConsuming` method. I forked your repo and created a [branch](https://github.com/fqshopify/fs2Graceful/compare/main...LMnet:fs2Graceful:ce3-fix) with a fix. I have combined reworked `uncancelable` functionality with the manual `fiber` manipulations. The final code looks easier and it's more compact.

I'm still need to update docs with the new example, though.

**Comment by fqshopify** [comment]
ha! you reached the same conclusion I did over the weekend, I also found I could get it to work on CE3 with `.start.flatMap(_.joinWithNever)` (see [PR here](https://github.com/fqshopify/fs2Graceful/pull/1/files) for more details) although I was still trying to understand why that worked XD Thanks for your explanation, it makes a lot of sense and your updated solution looks very nice. 

I will leave this ticket in your capable hands (although do let me know if there's anything I can do to help).
